lib/Transforms/SecretWithMgmt/SecretWithMgmtBGV.cpp

```diff
@@ -0,0 +1,385 @@
+        levelMap[blockArg] = 0;
+        mulMap[blockArg] = false;
+      }
+
+      genericOp.getBody()->walk<WalkOrder::PreOrder>([&](Operation *op) {
+        auto levelResult = 0;
+        for (auto operand : op->getOperands()) {
+          if (levelMap.count(operand) == 0) {
+            continue;
+          }
+          auto levelOperand = levelMap.at(operand);
+          levelResult = std::max(levelResult, levelOperand);
+        }
+
+        bool operandsMul = false;
+        for (auto operand : op->getOperands()) {
+          if (mulMap.count(operand) == 0) {
+            continue;
+          }
+          operandsMul |= mulMap.at(operand);
+        }
+
+        llvm::TypeSwitch<Operation &>(*op)
+            .Case<arith::MulIOp, secret::YieldOp, tensor::ExtractOp>(
+                [&](auto mulOp) {
+                  // mod reduced
+                  if ((isa<arith::MulIOp>(mulOp) ||
+                       isa<tensor::ExtractOp>(mulOp)) &&
+                      (includeFirst || operandsMul)) {
+                    levelResult += 1;
+                  }
+                  if (isa<secret::YieldOp>(mulOp) && operandsMul) {
+                    // mod reduce before yield if muled
+                    levelResult += 1;
+                  }
+                  // avoid yield op
+                  if (mulOp->getNumResults() != 0) {
+                    levelMap[mulOp->getResult(0)] = levelResult;
+                    mulMap[mulOp->getResult(0)] = true;
+                  }
+
+                  for (auto operand : mulOp->getOperands()) {
+                    auto secretness = getSecretness(operand);
+                    if (!secretness) {
+                      continue;
+                    }
+                    b.setInsertionPoint(mulOp);
+                    Value managed = operand;
+                    for (auto i = 0; i != levelResult - levelMap.at(operand);
+                         ++i) {
+                      managed =
+                          b.create<mgmt::ModReduceOp>(mulOp->getLoc(), managed);
+                      levelMap[managed] = levelMap.at(operand) + i + 1;
+                      mulMap[managed] = mulMap.at(operand);
+                    }
+                    mulOp->replaceUsesOfWith(operand, managed);
+                  }
+                })
+            .Default([&](auto &op) {
+              for (auto result : op.getResults()) {
+                levelMap[result] = levelResult;
+                mulMap[result] = operandsMul;
+              }
+            });
+      });
+    });
+  }
+
+  void alwaysModreduceWhenLevelMismatch() {
+    DataFlowSolver solver;
+    solver.load<dataflow::DeadCodeAnalysis>();
+    solver.load<dataflow::SparseConstantPropagation>();
+    solver.load<SecretnessAnalysis>();
+    if (failed(solver.initializeAndRun(getOperation()))) {
+      getOperation()->emitOpError() << "Failed to run the analysis.\n";
+      signalPassFailure();
+      return;
+    }
```

Comment:
This can probably be refactored into a function since you do it so often:

```cpp
LogicalResult runSecretnessAnalysis(DataFlowSolver &solver, Operation *operation) {
    solver.load<dataflow::DeadCodeAnalysis>();
    solver.load<dataflow::SparseConstantPropagation>();
    solver.load<SecretnessAnalysis>();
    if (failed(solver.initializeAndRun(getOperation()))) {
      operation->emitOpError() << "Failed to run the analysis.\n";
      return failure();
    }
}
```

Then

```cpp
DataFlowSolver solver;
if (failed(runSecretnessAnalysis(solver, getOperation()))) {
  signalPassFailure();
  return;
}
```